src/include/platform/DiagnosticDataProvider.h

```diff
@@ -0,0 +1,179 @@
+
+// Maximum length of vendor defined name or prefix of the software thread that is
+// static for the duration of the thread.
+static constexpr size_t kMaxThreadNameLength = 32;
+
+struct ThreadMetrics : public app::Clusters::SoftwareDiagnostics::Structs::ThreadMetrics::Type
+{
+    char NameBuf[kMaxThreadNameLength + 1];
+    ThreadMetrics * Next; /* Pointer to the next structure.  */
+};
+
+/**
+ * Defines the delegate class of Platform Manager to notify platform updates.
+ */
+class DiagnosticsDelegate
+{
+public:
+    virtual ~DiagnosticsDelegate() {}
+
+    /**
+     * @brief
+     *   Called after the current device is rebooted
+     */
+    virtual void OnDeviceRebooted() {}
+};
+
+/**
+ * Provides access to runtime and build-time configuration information for a chip device.
+ */
+class DiagnosticDataProvider
+{
+public:
+    void SetDelegate(DiagnosticsDelegate * delegate) { mDelegate = delegate; }
+    DiagnosticsDelegate * GetDelegate() const { return mDelegate; }
+
+    /**
+     * General Diagnostics methods.
+     */
+    virtual CHIP_ERROR GetRebootCount(uint16_t & rebootCount);
+    virtual CHIP_ERROR GetUpTime(uint64_t & upTime);
+    virtual CHIP_ERROR GetTotalOperationalHours(uint32_t & totalOperationalHours);
+    virtual CHIP_ERROR GetBootReasons(uint8_t & bootReasons);
+    virtual CHIP_ERROR GetActiveHardwareFaults(GeneralFaults<kMaxHardwareFaults> & hardwareFaults);
+    virtual CHIP_ERROR GetActiveRadioFaults(GeneralFaults<kMaxRadioFaults> & radioFaults);
+    virtual CHIP_ERROR GetActiveNetworkFaults(GeneralFaults<kMaxNetworkFaults> & networkFaults);
+
+    /**
+     * Software Diagnostics methods.
+     */
+    virtual CHIP_ERROR GetCurrentHeapFree(uint64_t & currentHeapFree);
+    virtual CHIP_ERROR GetCurrentHeapUsed(uint64_t & currentHeapUsed);
+    virtual CHIP_ERROR GetCurrentHeapHighWatermark(uint64_t & currentHeapHighWatermark);
+
+    /*
+     * Get the linked list of thread metrics of the current plaform. After usage, each caller of GetThreadMetrics
+     * needs to release the thread metrics list it gets via ReleaseThreadMetrics.
+     *
+     */
+    virtual CHIP_ERROR GetThreadMetrics(ThreadMetrics ** threadMetricsOut);
+    virtual void ReleaseThreadMetrics(ThreadMetrics * threadMetrics);
+
+protected:
+    // Construction/destruction limited to subclasses.
+    DiagnosticDataProvider()          = default;
+    virtual ~DiagnosticDataProvider() = default;
+
+private:
+    DiagnosticsDelegate * mDelegate = nullptr;
+
+    // No copy, move or assignment.
+    DiagnosticDataProvider(const DiagnosticDataProvider &)  = delete;
+    DiagnosticDataProvider(const DiagnosticDataProvider &&) = delete;
+    DiagnosticDataProvider & operator=(const DiagnosticDataProvider &) = delete;
+};
+
+/**
+ * Returns a reference to a DiagnosticDataProvider object.
+ *
+ * Applications should use this to access the features of the DiagnosticDataProvider.
+ */
+extern DiagnosticDataProvider & DiagnosticDataMgr();
```

Comment:
This does not need to say `extern`.

Also, it would be good for symettry to call this `SetDiagnosticDataProvider()`

Reply:
'extern' is optional, we can remove, just align the style with other platform interface classes.

We need to use different name than constructor name, just as use PlatformMgr for PlatformManager class.

Reply:
> We need to use different name than constructor name, just as use PlatformMgr for PlatformManager class.

I do not understand this comment.

The naming here is confusing. The class is called `DiagnosticsDataProvider`. Why is the getter for a singleton isntance called `DiagnosticsDataMgr` ?

Reply:
Suppose we have following singleton class which use same name for constructor and getter.

class Singleton {
public:
    static Singleton * Singleton(); // returns the unique instance
    static void destroyInstance(); // returns the unique instance
    void print();
    
private:
    static Singleton *instance;
    Singleton(); // Private constructor: a singleton instance is accessible only through instance
    ~Singleton();
    Singleton(const Singleton&); // prevent compiler-generated copy constructor
    Singleton& operator=(const Singleton&); // prevent compiler-generated operator =
};

I will get the following compile error:

singleton.cpp:11:34: error: return type specification for constructor invalid
   11 |     static Singleton * Singleton(); // returns the unique instance